File: Creational_Patterns/Singleton/python/DataBaseConnection.py
import threading
class DataBaseConnection:
  _instance = None
  _lock = threading.Lock()

  def __init__(self):
    if DataBaseConnection._instance is not None:
      raise Exception("Instance Exited")
    DataBaseConnection._instance = self

  @classmethod
  def get_instance(cls):
    if cls._instance is None:
      with cls._lock:
        if cls._instance is None:
          cls._instance = DataBaseConnection()
    return cls._instance
  
  def query(self,sql : str):
    print("excuting query...",sql)


ob1 = DataBaseConnection.get_instance()
ob2  = DataBaseConnection.get_instance()

print(ob1 is ob2)

ob1.query("select * from table")

File: Creational_Patterns/Singleton/Java/DataBaseConnection.java
class DataBaseConnection{
  private DataBaseConnection(){}

  private class Holder{
    private static final DataBaseConnection INSTANCE = new DataBaseConnection();
  }

  public static DataBaseConnection getInstance(){
    return Holder.INSTANCE;
  }

  public void query(String sql){
    System.out.println("Executing Query...."+ sql);
  }
}

class Main{
  public static void main(String[] args){
    DataBaseConnection ob1 = DataBaseConnection.getInstance();
    DataBaseConnection ob2 = DataBaseConnection.getInstance();

    System.out.println(ob1.equals(ob2));

    ob1.query("select * from table");
  }
}

